File: src/main/java/mods/cybercat/gigeresque/common/block/CustomSkullBlock.java
package mods.cybercat.gigeresque.common.block;

import net.minecraft.core.BlockPos;
import net.minecraft.util.Mth;
import net.minecraft.world.item.context.BlockPlaceContext;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.block.*;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.state.StateDefinition;
import net.minecraft.world.level.block.state.properties.BlockStateProperties;
import net.minecraft.world.level.block.state.properties.IntegerProperty;
import net.minecraft.world.phys.shapes.CollisionContext;
import net.minecraft.world.phys.shapes.Shapes;
import net.minecraft.world.phys.shapes.VoxelShape;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class CustomSkullBlock extends AbstractSkullBlock {
    public static final IntegerProperty ROTATION = BlockStateProperties.ROTATION_16;
    public static final VoxelShape SHAPE = box(4.0, 0.0, 4.0, 12.0, 8.0, 12.0);

    public CustomSkullBlock(SkullBlock.Type type, Properties settings) {
        super(type, settings);
        registerDefaultState(getStateDefinition().any().setValue(ROTATION, 0));
    }

    @Override
    public @NotNull VoxelShape getShape(@NotNull BlockState state, @NotNull BlockGetter world, @NotNull BlockPos pos, @NotNull CollisionContext context) {
        return SHAPE;
    }

    @Override
    public @NotNull VoxelShape getOcclusionShape(@NotNull BlockState state, @NotNull BlockGetter world, @NotNull BlockPos pos) {
        return Shapes.empty();
    }

    @Nullable
    @Override
    public BlockState getStateForPlacement(BlockPlaceContext ctx) {
        return defaultBlockState().setValue(ROTATION, Mth.floor((ctx.getRotation() * 16.0f / 360.0f) + 0.5) & 15);
    }

    @Override
    public @NotNull BlockState rotate(BlockState state, Rotation rotation) {
        return state.setValue(ROTATION, rotation.rotate(state.getValue(ROTATION), 16));
    }

    @Override
    public @NotNull BlockState mirror(BlockState state, Mirror mirror) {
        return state.setValue(ROTATION, mirror.mirror(state.getValue(ROTATION), 16));
    }

    @Override
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
        builder.add(ROTATION);
    }

    public enum AlienSkullType implements SkullBlock.Type {
        AQUA, CLASSIC, RUNNER,
    }
}

File: src/main/kotlin/com/bvanseg/gigeresque/common/block/CustomSkullBlock.kt
//package com.bvanseg.gigeresque.common.block
//
//import net.minecraft.block.*
//import net.minecraft.item.ItemPlacementContext
//import net.minecraft.state.StateManager
//import net.minecraft.state.property.IntProperty
//import net.minecraft.state.property.Properties
//import net.minecraft.util.BlockMirror
//import net.minecraft.util.BlockRotation
//import net.minecraft.util.math.BlockPos
//import net.minecraft.util.math.MathHelper
//import net.minecraft.util.shape.VoxelShape
//import net.minecraft.util.shape.VoxelShapes
//import net.minecraft.world.BlockView
//
///**
// * @author Boston Vanseghi
// */
//class CustomSkullBlock(skullType: AlienSkullType, settings: Settings) : AbstractSkullBlock(skullType, settings) {
//
//    companion object {
//        val ROTATION: IntProperty = Properties.ROTATION
//        val SHAPE: VoxelShape = createCuboidShape(4.0, 0.0, 4.0, 12.0, 8.0, 12.0)
//    }
//
//    enum class AlienSkullType : SkullBlock.SkullType {
//        AQUA,
//        CLASSIC,
//        RUNNER,
//    }
//
//    init {
//        defaultState = (stateManager.defaultState as BlockState).with(ROTATION, 0) as BlockState
//    }
//
//    override fun getOutlineShape(
//        state: BlockState,
//        world: BlockView,
//        pos: BlockPos,
//        context: ShapeContext
//    ): VoxelShape {
//        return SHAPE
//    }
//
//    override fun getCullingShape(state: BlockState, world: BlockView, pos: BlockPos): VoxelShape {
//        return VoxelShapes.empty()
//    }
//
//    override fun getPlacementState(ctx: ItemPlacementContext): BlockState {
//        return defaultState.with(
//            ROTATION,
//            MathHelper.floor((ctx.playerYaw * 16.0f / 360.0f).toDouble() + 0.5) and 15
//        ) as BlockState
//    }
//
//    override fun rotate(state: BlockState, rotation: BlockRotation): BlockState {
//        return state.with(ROTATION, rotation.rotate((state.get(ROTATION) as Int), 16)) as BlockState
//    }
//
//    override fun mirror(state: BlockState, mirror: BlockMirror): BlockState {
//        return state.with(ROTATION, mirror.mirror((state.get(ROTATION) as Int), 16)) as BlockState
//    }
//
//    override fun appendProperties(builder: StateManager.Builder<Block, BlockState>) {
//        builder.add(ROTATION)
//    }
//}